### PR TITLE
[WIP] Replace network plugin API with CNI

### PR DESCRIPTION
**Pull request by Symmetric**
A prototype implementation of Kubernetes network plugins using the [CNI](https://github.com/appc/cni).

This provides a richer interface to plugins than the current exec-API, while remaining un-opinionated and flexible on the subject of how plugins should implement connectivity between containers.

Need to figure out how to handle the migration from the current exec-API. Some options:

i) Cut over and offer no migration path; every plugin implementor must re-implement to the CNI
ii) Build a CNI plugin that invokes old exec-API plugins, so that existing implementors can call kubelet with --net-plugin=legacy:pluginName (equivalent to calling --net-plugin=pluginName on the current API)
iii) Keep the current plugin code in-tree, so that exec-API plugins are not broken by adding this code; deprecate the exec-API, and remove it at a later date.

cc @thockin @lxpollitt @caseydavenport 

TODO:
- [ ] Tests
- [ ] Migration plan from current plugin API
- [ ] Verify correct handling of error case when pod is created, but sync of pod.Status.podIP to APIServer fails.
- [ ] Update cni Godeps once @squaremo has upstreamed his libcni branch (which this PR depends on).

### CODE REVIEW

**Comment by Symmetric** [comment]
cc @squaremo

**Comment by ncdc** [comment]
@kubernetes/rh-networking

**Comment by rajatchopra** [comment]
@Symmetric Good work here. Thanks for writing this up for everyone to get a feel of the new way.

For the migration from existing plugins, there are two parts to the issue:
1. Note that the existing plugins are not just exec based. Exec is just one of the plugins (though the only one in the master tree today). One can compile in plugins too (OpenShift does that) using the current API model.
2. Even for the exec based plugin, I would vote to not go for option (i), that is no 'hard cut over' please. There are at least 5 efforts using it that I am aware of and some of the code is used in production too. Option (ii) is a good idea - create a CNI plugin that emulates the old style exec plugin.

On this implementation I feel we should also discuss how to have 'multiple networks'.

**Comment by Symmetric** [comment]
@rajatchopra we briefly discussed compiled-in plugins on [SIG-network](https://groups.google.com/d/msg/kubernetes-sig-network/5MWRPxsURUw/K799JL6KBQAJ), but I didn't manage to collect any feedback either way. Do you think in-tree plugins are a hard requirement, or could vendors like OpenShift live with only the CNI?

The switch to CNI makes one significant change, namely that the plugin now returns the IP address, and it's up to the orchestrator (k8s) to keep track of it since there's no STATUS call to get it later. This probably necessitates changes to the existing in-process plugin API anyway, so maybe standardizing on CNI would be cleanest. I don't have a feel for what the in-tree plugins require though (if it's more than this proposal offers, then perhaps that shows that this implementation is inadequate).

Regarding multiple networks, I think we can extend the current API cleanly to get there; CNI supports multiple networks, and will create one veth per joined network in the container namespace. Since Networks are just simple config blobs, they can be mapped to API objects if necessary. Pods can join any number of Networks, in principle.

In short I think this proposal sets us up nicely for multiple networks/IPs per pod.

**Comment by rajatchopra** [comment]
@Symmetric Agree. Here is an alternate implementation that implements CNI as a kube plugin itself. 

https://github.com/kubernetes/kubernetes/pull/13766 - uses the libcni from @squaremo and code from this PR, but retains the possibility of compiling a plugin in.

We should discuss the Status hook magic still, not sure if storing the value returned by AddNetwork is good enough forever.

**Comment by dcbw** [comment]
@Symmetric it seems like the network plugin name given to kubernetes is no longer the same thing as it was before.  eg before, it was actually a plugin name, while now it's the name of a config file in /etc/cni/net.d that looks like https://github.com/appc/cni#running-the-plugins .  I think that's confusing because the only thing that's useful in that config file is the 'type' element which tells us which plugin/driver to use.

It gets more confusing with multi-network features like we implement in OpenShift, where different pods could be part of any number of different networks; so while Kubernetes here reads only one network file, it would be using that file/config for *multiple* different isolated networks internally, depending on that driver.  That seems rather odd.

I would rather have Kubernetes not use the network plugin name as an index into the /etc/cni/net.d files' 'name' fields, but instead use it as the actual plugin name (eg, 'type').  The network name used to find the specific net.d config file would come from something else in Kubernetes, like the 'namespace' name, or some other indicator of isolation that we all agree on here.  Just not the plugin name -> network name, since that can't change after k8s runs, but the network a pod is put in may well change at runtime.

**Comment by Symmetric** [comment]
@dcbw good spot, I agree, the plugin name should select the CNI type, not network name.

Based on our SIG-network chat today, I don't think we want to implement Namespace isolation using different CNI networks. I do agree that whatever mechanism that we choose for multiple network interfaces per pod should be the thing that enables multiple CNI networks.

For now I propose that we make the change that you suggested to have plugin name select the CNI type, and decide multiple interfaces in a subsequent PR.

**Comment by Symmetric** [comment]
@rajatchopra following on from the SIG-network hangout today, can we recap?

1) We don't want to cut over to CNI without a migration path. I prefer option ii) above (CNI->exec-plugin bridge), though it does have issues with INIT and STATUS actions that would need to be resolved. As a backup plan we could keep the old exec interface, as your PR suggests, but that's a much less clean cut-over.
2) I think we agreed that we could migrate away from compiled-in plugins. Are you able to do that in the 1.1 timeframe, or do you want to keep the compiled-in interface in this iteration?
3) As I'm on vacation for the next week, does anyone want to volunteer to progress the implementation of the plugin interface (be it on this PR or Rajat's version)?

cc @caseydavenport

**Comment by lxpollitt** [comment]
Note @squaremo PR: https://github.com/appc/cni/pull/59

**Comment by rajatchopra** [comment]
> 1) We don't want to cut over to CNI without a migration path. I prefer option ii) above (CNI->exec->
> plugin bridge), though it does have issues with INIT and STATUS actions that would need to be 
> resolved. As a backup plan we could keep the old exec interface, as your PR suggests, but that's 
> a much less clean cut-over.
> 2) I think we agreed that we could migrate away from compiled-in plugins. Are you able to do that in > the 1.1 timeframe, or do you want to keep the compiled-in interface in this iteration?

I am afraid we need the compiled-in plugins to be kept intact for the time being. I agree that my version of CNI support is not the cleanest cut-over, but despite the intent I could not come up with something that will allow the current ones to work as well as get the CNI one working.

@jainvipin ^^, agree?


> 3) As I'm on vacation for the next week, does anyone want to volunteer to progress the implementation of the plugin interface (be it on this PR or Rajat's version)?

Happy to take it forward with my PR if everyone agrees with that approach. Referencing again: #13766

**Comment by lxpollitt** [comment]
@thockin I think the direction we go would benefit from your blessing. 

@rajatchopra, your offer to run with taking this forward in time for 1,1 is very much appreciated. @thockinh please jump in if you already have someone at Google already lined up to do this.

**Comment by lxpollitt** [comment]
(Note I am traveling, currently in London. So time zine wise I won't be able to respond to further comments until tomorrow.)

**Comment by jainvipin** [comment]
> >  I am afraid we need the compiled-in plugins to be kept intact for the time being. I agree that my version of CNI support is not the cleanest cut-over, but despite the intent I could not come up with something that will allow the current ones to work as well as get the CNI one working.

> @jainvipin ^^, agree?
While I agree with your assesment (about cleanest cut-over), I do not use compiled-in plugin. Therefore, as an exec based plugins user, a transition path is a critical requirement for me.

**Comment by caseydavenport** [comment]
It looks like we can say the following in the v1.1 timeframe:
  - We need to maintain support for old-style plugins
  - We need to maintain support for exec-style and compiled-in plugins.

Option iii (Maintain both plugin types in-tree, deprecate exec plugin) seems like most achievable way to do this in the v1.1 time-frame, and gives time for plugin implementors to make the switch.  It doesn't prevent us from adding a CNI -> Exec mapping in the future, but gives us a bit more time to make an architecturally correct decision.

@thockin @lxpollitt @jainvipin @rajatchopra  - Does this line of thinking make sense? Would it meet our requirements for v1.1?  I'm wary that the 1.1 code-freeze date is fast approaching.  Does it make sense to have a mini-hangout to discuss our options here?

**Comment by caseydavenport** [comment]
@rajatchopra - In the absence of any chatter here, maybe you should drive forward with your PR?

**Comment by jainvipin** [comment]
@caseydavenport - works for me 

>  It looks like we can say the following in the v1.1 timeframe:
>
>    We need to maintain support for old-style plugins

>    We need to maintain support for exec-style and compiled-in plugins.

> Option iii (Maintain both plugin types in-tree, deprecate exec plugin) seems like most achievable way to do this in the v1.1 time-frame, and gives time for plugin implementors to make the switch. It doesn't prevent us from adding a CNI -> Exec mapping in the future, but gives us a bit more time to make an architecturally correct decision.

> @thockin @lxpollitt @jainvipin @rajatchopra - Does this line of thinking make sense? Would it meet our requirements for v1.1? I'm wary that the 1.1 code-freeze date is fast approaching. Does it make sense to have a mini-hangout to discuss our options here?

**Comment by dcbw** [comment]
Should we close this in favor of @rajatchopra's PR?  It seems like we've got two very similar issues and if @rajatchopra is delegated to continue the work we could combine things...

**Comment by thockin** [comment]
As much as I love all the red in this PR (deleted lines), I think we NEED to have a compat story for a release or two.  Given that, I think @rajatchopra's approach has better legs than this one.

Paul - thank you for spearheading this effort, though :)
